### PR TITLE
Custom exceptions

### PR DESCRIPTION
**Pull request by kiran-chaudhary**
### Description of Change ###

Added custom exceptions in detection lists `remove` method.
  
### Issues Resolved ###

- closes #1189

### Testing Procedure ###
```
try:
 sdk.detectionlists.departingemployee.remove("xyz") 
 sdk.detectionlists.highriskemployee.remove("xyzzy")
except Py42UserDoesNotExistError:
```
### PR Checklist ###
Did you remember to do the below?

- [x] Add unit tests to verify this change
- [x] Add an entry to CHANGELOG.md describing this change
- [x] Add docstrings for any new public parameters / methods / classes

### CODE REVIEW

**Comment by alanag13** [comment]
@kiran-chaudhary if the request returns a 400, then thew new exception type should subclass `Py42BadRequestError` instead of `NotFound` so that users who are catching this exception will have it continue to be caught. Thanks!

File: src/py42/exceptions.py [hunk 1]
```diff
@@ -167,6 +167,13 @@ def __init__(self, exception, message=None):
         super(Py42UserAlreadyExistsError, self).__init__(exception, message)
 
 
+class Py42UserDoesNotExist(Py42NotFoundError):
```
Comment (unparalleled-js) [inline src/py42/exceptions.py:170]: I think the name of the exception is misleading. It makes it sound like either the user does not exist in Code42 or the user does not have a detection list profile.

A user is able to have a detection list profile and still not be on a detection list.

That makes it likely to get confused with Code42 CLI exception `UserDoesNotExistError`.

We should rename this to be something like `Py42UserNotOnListError`.

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,10 @@ how a consumer would use the library (e.g. adding unit tests, updating documenta
 
 - Bug where an empty destination list in a device's backup set broke creation of DeviceSettings objects for that device.
 
+### Added
+
+- Added new exception `Py42UserDoesNotExist` to throw error if `remove()` throws `400` when user does not exist in the detection list.
```
Comment (unparalleled-js) [inline CHANGELOG.md:19]: We should call this exception something else

File: src/py42/services/detectionlists/high_risk_employee.py [hunk 3]
```diff
@@ -91,7 +93,14 @@ def remove(self, user_id):
             u"userId": user_id,
         }
         uri = self._make_uri(u"/remove")
-        return self._connection.post(uri, json=data)
+        try:
+            return self._connection.post(uri, json=data)
+        except Py42NotFoundError as err:
+            message = (
+                "User {} is not currently on the high-risk-employee "
```
Comment (unparalleled-js) [inline src/py42/services/detectionlists/high_risk_employee.py:100]: Would be better if it were `User with ID '{}' is not ... `

That way, it matches the other exceptions and reads more truthfully

File: src/py42/services/detectionlists/departing_employee.py [hunk 4]
```diff
@@ -97,7 +99,14 @@ def remove(self, user_id):
         tenant_id = self._user_context.get_current_tenant_id()
         uri = self._uri_prefix.format(u"remove")
         data = {u"userId": user_id, u"tenantId": tenant_id}
-        return self._connection.post(uri, json=data)
+        try:
+            return self._connection.post(uri, json=data)
+        except Py42NotFoundError as err:
+            message = (
+                "User {} is not currently on the departing-employee "
```
Comment (unparalleled-js) [inline src/py42/services/detectionlists/departing_employee.py:106]: Should be `User with ID `{}` ... `

File: CHANGELOG.md (hunk 2 above)
Comment (unparalleled-js) [inline CHANGELOG.md:19]: It is actually 404 I thought

File: tests/services/detectionlists/test_departing_employee.py [hunk 5]
```diff
@@ -107,6 +108,18 @@ def mock_py42_response(self, mocker, mock_get_case_details_function, py42_respon
...
 
+    @pytest.fixture
+    def mock_detection_list_user_client_failure(
+        self, mock_connection, user_context, mock_user_client, mocker
+    ):
+        user_client = DetectionListUserService(
+            mock_connection, user_context, mock_user_client
+        )
+        response = mocker.MagicMock(spec=Response)
+        response.status_code = 400
```
Comment (unparalleled-js) [inline tests/services/detectionlists/test_departing_employee.py:119]: Technically this should be 404 I thought?

Is this test passing?

File: tests/services/detectionlists/test_departing_employee.py [hunk 6]
```diff
@@ -107,6 +108,18 @@ def mock_py42_response(self, mocker, mock_get_case_details_function, py42_respon
 
         return py42_response
 
+    @pytest.fixture
+    def mock_detection_list_user_client_failure(
```
Comment (unparalleled-js) [inline tests/services/detectionlists/test_departing_employee.py:112]: What kind of failure?

Should be `mock_detection_list_user_not_found_client`

File: tests/conftest.py [hunk 7]
```diff
@@ -166,3 +167,13 @@ def mock_connection(mocker):
 def mock_successful_connection(mock_connection, successful_response):
     mock_connection.get.return_value = successful_response
     return mock_connection
+
+
+@pytest.fixture
+def mock_connection_post_failure(mocker, mock_connection):
```
Comment (unparalleled-js) [inline tests/conftest.py:173]: This fixture does not say enough about what it is... Should be

`mock_post_not_found_session` or something

**Review by unparalleled-js — CHANGES_REQUESTED** [review]
Not quite ready yet - see comments

File: src/py42/exceptions.py [hunk 8]
```diff
@@ -167,6 +167,13 @@ def __init__(self, exception, message=None):
...
+    """An exception raised when user does not exist in the detection list."""
```
Comment (unparalleled-js) [inline src/py42/exceptions.py:171]: I'd also modify the doc to say `... when user is not in the detection list` rather than `not exist`

**Comment by unparalleled-js** [comment]
> @kiran-chaudhary if the request returns a 400, then the new exception type should subclass `Py42BadRequestError` instead of `NotFound` so that users who are catching this exception will have it continue to be caught. Thanks!

It's actually supposed to be 404

File: tests/conftest.py (hunk 7 above)
Comment (kiran-chaudhary) [reply tests/conftest.py]: Renamed.

File: tests/services/detectionlists/test_departing_employee.py (hunk 6 above)
Comment (kiran-chaudhary) [reply tests/services/detectionlists/test_departing_employee.py]: Removed.

File: CHANGELOG.md (hunk 2 above)
Comment (kiran-chaudhary) [reply CHANGELOG.md]: Changed.

File: src/py42/services/detectionlists/departing_employee.py (hunk 4 above)
Comment (kiran-chaudhary) [reply src/py42/services/detectionlists/departing_employee.py]: Changed.

File: src/py42/services/detectionlists/high_risk_employee.py (hunk 3 above)
Comment (kiran-chaudhary) [reply src/py42/services/detectionlists/high_risk_employee.py]: Changed.

File: CHANGELOG.md (hunk 2 above)
Comment (kiran-chaudhary) [reply CHANGELOG.md]: Changed.

File: src/py42/exceptions.py (hunk 1 above)
Comment (kiran-chaudhary) [reply src/py42/exceptions.py]: Changed.

File: src/py42/exceptions.py (hunk 8 above)
Comment (kiran-chaudhary) [reply src/py42/exceptions.py]: changed.

File: src/py42/exceptions.py [hunk 9]
```diff
@@ -167,8 +167,8 @@ def __init__(self, exception, message=None):
         super(Py42UserAlreadyExistsError, self).__init__(exception, message)
 
 
-class Py42UserDoesNotExist(Py42NotFoundError):
-    """An exception raised when user does not exist in the detection list."""
+class Py42UserNotOnListError(Py42NotFoundError):
+    """An exception raised when user is not in the detection list."""
```
Comment (unparalleled-js) [inline src/py42/exceptions.py:171]: the user

Comment (unparalleled-js) [reply src/py42/exceptions.py:171]: on the detection list, instead of `in`

**Review by unparalleled-js — APPROVED** [review]

**Pull request merged**